db/commitlog/commitlog.cc

```diff
@@ -1392,83 +1393,81 @@ void db::commitlog::segment_manager::flush_segments(uint64_t size_to_remove) {
-                        return repeat([this, f, &rem, &buf]() mutable {
-                            if (rem == 0) {
-                                return make_ready_future<stop_iteration>(stop_iteration::yes);
-                            }
-                            static constexpr size_t max_write = 128 * 1024;
-                            auto n = std::min(max_write / buf_size, 1 + rem / buf_size);
-
-                            std::vector<iovec> v;
-                            v.reserve(n);
-                            size_t m = 0;
-                            while (m < rem && n < max_write) {
-                                auto s = std::min(rem - m, buf_size);
-                                v.emplace_back(iovec{ buf.get_write(), s});
-                                m += s;
-                            }
-                            return f.dma_write(max_size - rem, std::move(v), service::get_local_commitlog_priority()).then([&rem](size_t s) mutable {
-                                rem -= s;
-                                return stop_iteration::no;
-                            });
-                        });
-                    });
-                });
-            });
+
+                co_await f.allocate(existing_size, max_size - existing_size);
+
+                static constexpr size_t buf_size = 4 * segment::alignment;
+                size_t zerofill_size = max_size - align_down(existing_size, segment::alignment);
+                auto rem = zerofill_size;
+
+                auto buf = allocate_single_buffer(buf_size);
+                while (rem != 0) {
+                    static constexpr size_t max_write = 128 * 1024;
+                    auto n = std::min(max_write / buf_size, 1 + rem / buf_size);
+
+                    std::vector<iovec> v;
+                    v.reserve(n);
+                    size_t m = 0;
+                    while (m < rem && n < max_write) {
+                        auto s = std::min(rem - m, buf_size);
+                        v.emplace_back(iovec{ buf.get_write(), s});
+                        m += s;
+                    }
+                    auto s = co_await f.dma_write(max_size - rem, std::move(v), service::get_local_commitlog_priority());
+                    rem -= s;
+                }
+
+                // sync metadata (size/written)
+                co_await f.flush();
+            }
         } else {
-            fut = f.truncate(max_size);
+            co_await f.truncate(max_size);
         }
-        return fut.then([this, d, f, filename] () mutable {
-            auto s = make_shared<segment>(shared_from_this(), std::move(d), std::move(f), max_size);
-            return make_ready_future<sseg_ptr>(s);
-        });
-    });
+
+        if (cfg.extensions && !cfg.extensions->commitlog_file_extensions().empty()) {
+            for (auto * ext : cfg.extensions->commitlog_file_extensions()) {
+                auto nf = co_await ext->wrap_file(std::move(filename), f, flags);
```

Comment:
No, it does not. The old code would only close the contents of `fut` - i.e. the `f` file variable on exception. This content can mutate in the loop, yes, but it is extensions responsibility to either close or wrap-for-subsequent-close if creating any secondary file. 
The exact level of protection is ensured here. If a co_await throws, we will end up in exception handler below, with f being the last file opened, and close it subsequently.

Reply:
No, the error is propagated up the continuations chain and causes a domino effect of closing all intermediate files.
See https://github.com/bhalevy/seastar/commit/aff4490c52ca50c2f5da2ed5d7dd637cf6692fe6

Reply:
No, it does not. It closes the exact single file put into it, then propagates the exception. In the unmodified allocate.. code, each call into an extension is wrapped in a `with_file_close_on_failure`, yes. This is the equivalent of wrapping it in a try catch that closes the _exact_ var use to call into the extension. Nothing more, nothing less. The same "chaining" is accomplished by the co_await calls. As long as nothing fails, any closing of discarded files must still be done by extension itself. And if exception happens, only the file at the exception is handled, just as above. Because all other parts of the chain must already have happened.

Reply:
OK, I read the extensions implementation and they indeed close their underlying files, so this will work.

LGTM now